src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/AsnWriterExtensions.cs

```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Formats.Asn1;
+using System.Text;
+
+namespace System.DirectoryServices.Protocols
+{
+    internal static class AsnWriterExtensions
+    {
+        public static void WriteLdapString(this AsnWriter writer, string value, Encoding stringEncoding, bool mandatory = true, Asn1Tag? tag = null)
```

Comment:
I'm not sure this is a good name.  There's no construct (that I see) called `LdapString`, and not all strings in LDAP are sent as "A Utf8String, except using tag 04 instead of 0C".

`WriteUtf8OctetString`, maybe?

The `bool mandatory` has no peer on AsnWriter methods.  I recommend removing it here (making it always behave as `true`, and making the one "optional" caller bring that logic closer to home... so it looks like any other conditional write for an ASN OPTIONAL or DEFAULT value.

Reply:
The name `LdapString` partially comes from [RFC2251](https://datatracker.ietf.org/doc/html/rfc2251#section-4.1.5), as the backing type for `AttributeDescription`. Do you still want the name to change?

It was primarily used for writing the sort controls, and the other control logic piggybacks on the same method by explicitly specifying the encoding. I'll see if two methods would be clearer for this.

Reply:
I've tried a couple of different methods to see what the semantics look like, and agree - `WriteStringAsOctetString` it is. That's rolled up and done now.